### PR TITLE
powershell rest msi changes

### PR DESCRIPTION
**Pull request by rokmicrosoft**

### CODE REVIEW

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 1]
```diff
@@ -377,6 +391,53 @@ function Get-SpnAccessToken {
     }
 }
 
+# Get the Bearer Access Token from the Endpoint
+function Get-MsiAccessToken {
+    [CmdletBinding()]
+    param([Parameter(Mandatory=$true)] $endpoint)
+
+    $msiPort = 50342;
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: This logic has to change.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 2]
```diff
@@ -377,6 +391,53 @@ function Get-SpnAccessToken {
...
+    # Prepare contents for GET
+    $method = "GET"
+    $authUri = "http://localhost:$msiPort/oauth2/token"
+    
+    # Call Rest API to fetch AccessToken
+    Write-Verbose "Fetching Access Token For MSI"
+    
+    try
+    {
+        $proxyUri = Get-ProxyUri $authUri
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: When the endpoint is a local host, do we need proxy?

File: Tasks/Common/VstsAzureRestHelpers_/module.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
         "AZ_NoValidResourceIdFound": "No resource found with serverName {0}, serverType {1} in subscription {2}. Specify the correct serverName/serverType and try again.",
         "AZ_UnableToFetchLoadbalancerDetails": "Unable to fetch load balancer details",
         "AZ_UnableToFetchPublicIPAddressesDetails": "Unable to fetch public IP Addresses details",
-        "AZ_UnableToFetchNetworkInterfacesDetails": "Unable to fetch network interfaces details"
+        "AZ_UnableToFetchNetworkInterfacesDetails": "Unable to fetch network interfaces details",
+        "AZ_MsiAccessTokenFetchFailure": "Could not fetch access token for Managed Service Principal. Please configure Managed Service Identity (MSI) for virtual machine 'https://aka.ms/azure-msi-docs'.{0}"
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/module.json]: Review the message with @RoopeshNair

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 (hunk 1 above)
Comment (RoopeshNair) [reply Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: We are switching to IMDS service: https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service 

IMDS is now generally available as per the doc link above.

Comment (rokmicrosoft) [reply Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Done. updated.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 4]
```diff
@@ -377,6 +391,84 @@ function Get-SpnAccessToken {
     }
 }
 
+# Get the Bearer Access Token from the Endpoint
+function Get-MsiAccessToken {
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Can we have single copy of this instead of duplicates for REST and commands.

Comment (rokmicrosoft) [reply Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Take this comment in separate PR.

**Pull request merged**